subprojects/client-protocol/src/main/java/org/gradle/profiler/client/protocol/serialization/MessageSerializer.java

```diff
@@ -0,0 +1,82 @@
+package org.gradle.profiler.client.protocol.serialization;
+
+import org.gradle.profiler.client.protocol.Connection;
+import org.gradle.profiler.client.protocol.messages.*;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+public class MessageSerializer {
+    private static final Object NULL = new Object();
+    private final String peerName;
+    private final Connection connection;
+
+    public MessageSerializer(String peerName, Connection connection) {
+        this.peerName = peerName;
+        this.connection = connection;
+    }
+
+    public void send(Message message) {
+        try {
+            byte messageId = MessageRegistry.findMessageId(message.getClass());
+            connection.writeByte(messageId);
+            MessageReaderWriter readerWriter = MessageRegistry.findMessageReaderWriter(message.getClass());
+            readerWriter.writeTo(connection, message);
+            connection.flush();
+        } catch (IOException e) {
+            throw new RuntimeException(String.format("Could not write to %s.", peerName), e);
+        }
+    }
+
+    public <T extends Message> T receive(Class<T> type, Duration timeout) {
+        BlockingQueue<Object> queue = new LinkedBlockingQueue<>();
```

Comment:
Wouldn't it be simpler to submit a `Callable` to an `ExecuterService` here instead of starting a new thread and using a blocking queue?

Reply:
This is here in diff just to the renaming. But I refactored it anyway, it's much nicer! Thanks